### DISCUSSION TITLE
Filter for creating a Fence (none-straight slice)

### DISCUSSION THREAD

**Discussion by Jimmy-KL** [pyvista/pyvista#6154] [Ideas]
### Describe the feature you would like to be added.

There are several filters in Pyvsita for slicing, but there isn't a proper slice filter for creating a fence from a mesh. Fence is a terminology in geology that means the slice is not straight along a line while along a polyline.
There is a similar filter called [slice_slong_line](https://docs.pyvista.org/version/stable/api/core/_autosummary/pyvista.datasetfilters.slice_along_line), however, this filter will generate a slice across the whole mesh. Checking the example below, we will see that the slice line is only in part of the mesh, but the result is across the entire mesh. So if we want to generate a slice (fence) exactly along the polyline, this filter will fail. The fence filter can be very useful in geology to generate none-straight slice along boreholes.
![image](https://github.com/pyvista/pyvista/assets/114913030/d906dd41-d4c9-454b-b171-fc3473e54e01)

### Links to VTK Documentation, Examples, or Class Definitions.

I tried to find a similar thing in VTK, but I didn't find one.

### Pseudocode or Screenshots

I think we can use filter [slice_implicit](https://docs.pyvista.org/version/stable/api/core/_autosummary/pyvista.datasetfilters.slice_implicit) to create each straight part of the slice then merge them. But I didn't try and I think it's not a very efficient solution. Does anyone have a good idea? Or maybe there is a filter in VTK, I simply didn't find it.
```python
Insert code here
```
Or drag your screenshot here!

**Comment by user27182** [comment]
Is this what you mean?
![image](https://github.com/pyvista/pyvista/assets/89109579/9abec83e-d8a9-428c-9754-d300648ea3e8)

This result can be achieved by using the `extract_geometry` to basically clip the extents of the generated line to match the input line. E.g. we can add these lines to the [slice_slong_line](https://docs.pyvista.org/version/stable/api/core/_autosummary/pyvista.datasetfilters.slice_along_line) example:
``` python
# Only keep XY extents defined by the line
line_slice_extent = [*arc.bounds[:4],-np.inf, np.inf]
line_slice = line_slice.extract_geometry(extent=line_slice_extent)
```

Full code:

``` python
import numpy as np
import pyvista as pv
from pyvista import examples
hills = examples.load_random_hills()
center = np.array(hills.center)
point_a = center + np.array([5, 0, 0])
point_b = center + np.array([-5, 0, 0])
arc = pv.CircularArc(point_a, point_b, center, resolution=100)
line_slice = hills.slice_along_line(arc)

# Only keep XY extents defined by the line
line_slice_extent = [*arc.bounds[:4],-np.inf, np.inf]
line_slice = line_slice.extract_geometry(extent=line_slice_extent)

pl = pv.Plotter()
_ = pl.add_mesh(hills, smooth_shading=True, style='wireframe')
_ = pl.add_mesh(
    line_slice,
    line_width=10,
    render_lines_as_tubes=True,
    color='k',
)
_ = pl.add_mesh(arc, line_width=10, color='grey')
pl.show()
```

**Reply by Jimmy-KL** [reply]
Yes, this is what I mean - only the part that follows the polyline is needed. But I find 2 problems with this solution. 
1. The `extract_geometry` method will not split the cells at the ends of the polyline. In the following example I tried, it's the reason why the slice is not exactly following the polyline - two parts at both ends are missing.
![image](https://github.com/pyvista/pyvista/assets/114913030/a9f3e1e8-0cbe-434b-8d70-034741feaafc)
2. Another problem I can imagine is if the polyline is like the following shape. The black line is the polyline I will use to generate the slice, the red line is the line that will extend to the boundary of the mesh automatically. But because the x coordinate of the right end of the polyline is not the maximum x coordinate, when I try to extract the geometry, the red line connecting to the right end will be extracted too.
![image](https://github.com/pyvista/pyvista/assets/114913030/6ff311ed-617b-4156-ac34-8b0a32b4c15c)
I can imagine a solution that first generates a thin buff zone around the line, then clips the mesh using filter [clip_surface](https://docs.pyvista.org/version/stable/api/core/_autosummary/pyvista.datasetfilters.clip_surface).  With the 2 steps, we get a thin buff zone, then call filter [slice_slong_line](https://docs.pyvista.org/version/stable/api/core/_autosummary/pyvista.datasetfilters.slice_along_line). I think this solution can work out for most of the cases (if the mesh has coarse resolution, there may be some problem too), and it's not efficient when the mesh has high resolution.

**Reply by user27182** [reply]
> I can imagine a solution that first generates a thin buff zone around the line

You could use [`vtkConvexHull2D`](https://vtk.org/doc/nightly/html/classvtkConvexHull2D.html#details) on the XY coordinates of the input polyline to get a closed, convex "wrapping" representing the 2D bounds for clipping. And then extrude the convex XY coordinate up to z-min and z-max of the input surface. This should define the appropriate 3D bounding geometry of interest. 

And then use this to either clip the input surface before using [slice_slong_line](https://docs.pyvista.org/version/stable/api/core/_autosummary/pyvista.datasetfilters.slice_along_line) as you suggested, or alternatively use it to clip the generated line afterwards (as applied in my earlier example) - either or both might work, depending on how the clipping is applied.

**Reply by user27182** [reply]
How about this solution? The line is first converted to a ribbon to create the thin buffer zone around the line. And then it's extruded create a 3D bounding region to clip the input surface with. And then use `slice_along_line` on the clipped surface.
![image](https://github.com/pyvista/pyvista/assets/89109579/705e846e-2453-447e-89be-b91f98920450)

```python
import numpy as np
import pyvista as pv
from pyvista import examples

def clip_line(mesh,line, width=None):
    # Create ribbon from line
    width=line.length/10
    ribbon=line.ribbon(width=width)

    # Extrude ribbon from z_min to z_max
    ribbon.points[:,2] = mesh.bounds[4]
    z_magnitude=mesh.bounds[5]-mesh.bounds[4]
    ribbon=ribbon.extrude([0,0,z_magnitude], capping=True)
    ribbon=ribbon.compute_normals(auto_orient_normals=True)  # needed for clipping
    # Use extruded ribbon to clip the input surface
    return mesh.clip_surface(ribbon)

# Create mesh and line
mesh = examples.load_random_hills()
center = np.array(mesh.center)
point_a = center + np.array([5, 0, 0])
point_b = center + np.array([-5, 0, 0])
line = pv.CircularArc(point_a, point_b, center, resolution=100)

# Clip the mesh with the line
clipped = clip_line(mesh, line)

# Slice the clipped mesh with the line
fence = clipped.slice_along_line(line)

pl = pv.Plotter()
_ = pl.add_mesh(mesh, smooth_shading=True, style='wireframe')
_ = pl.add_mesh(
    fence,
    line_width=10,
    render_lines_as_tubes=True,
    color='k',
)
_ = pl.add_mesh(clipped,color='red', style='wireframe')
pl.show()
```

**Reply by Jimmy-KL** [reply]
Thanks for the code example. By the way, will slice first then clip be faster than clip first then slice? I haven't tested it yet. I have no idea of how VTK works at the low level.
